### DISCUSSION TITLE
GitHub Desktop for Mac deleted half of my project files.

### DISCUSSION THREAD

**Discussion by silver978** [community/community#23016] [New to GitHub]
<p>Good morning,</p>
<p>I am a beginner with Git and I am learning by using it for a project.</p>
<p>I use GitHub Desktop for Mac.</p>
<p>I recently committed some changes of my project, but I wanted to revert the commit with the purpose of removing the virtual commit but not the real files (so I would edit the project and generate a new commit).</p>
<p>So, I clicked the Undo button and I deleted the latest commit, but also all my files.</p>
<p>The file loss is big, but I mentally went through it in order to resolve the problem of my project in a better way.</p>
<p>The purpose of this post is: why the Undo removed also all my files? Is because I reverted to an old version of my project (to an old commit)? If so, how can I remove a virtual commit without deleting all my progress?</p>
<p>Probably I’ll switch directly to the console commands in order to have full control on my repo.</p>
<p>Thanks in advance!</p>

**Comment by steveward** [comment]
<p><a class="mention" href="/silver978">@silver978</a> sorry to hear you are having trouble! The <code>Undo</code> button in GitHub Desktop does not delete your files – it undoes the commit and then leaves the files associated with that commit as uncommitted changes in your working directory.</p>
<p>&gt;Is because I reverted to an old version of my project (to an old commit)?</p>
<p>Did you use the <code>revert</code> function in GitHub Desktop, or something else? If you attempt to revert to a previous commit with uncommitted changes in your working directory GitHub Desktop will throw an error, since you could lose those uncommitted changes.</p>
<p>The good news is that since you previously committed those files we should be able to recover any lost changes. Could you open up the command line (<code>Repository</code> &gt; <code>Open in \&lt;your_command_line_application\&gt;</code>) and run the following command?</p>
<p><code>git reflog</code></p>
<p>Please share the output of that command. It should give a history of what happened in your repository.</p>

**Comment by silver978** [comment]
<p>Thank you for your fast support!</p>
<p>As I see from the log, I used the revert function but without knowing much of it, my fault.</p>
<p>This is the output of the <code>git reflog</code> command:</p>
<pre><code>6e880ec (HEAD -&gt; master, origin/master, origin/HEAD) HEAD@{0}: reset: moving to 6e880ec2f19f35f5118d2eb693d0a14f68c32f47
d6c3e83 HEAD@{1}: reset: moving to d6c3e837bd7843fcdc778bcd1873b190fbc82a43
ef3f2c0 HEAD@{2}: revert: Revert "Added Map1 + Added collisions for Map1"
d6c3e83 HEAD@{3}: commit: Added Map1 + Added collisions for Map1
6e880ec (HEAD -&gt; master, origin/master, origin/HEAD) HEAD@{4}: commit: Alpha logo intro
d1a3886 HEAD@{5}: pull --progress origin: Fast-forward
f9f9294 HEAD@{6}: commit: Game juice in progress
576ee4e HEAD@{7}: commit: Prototype in progress
ac7d772 HEAD@{8}: clone: from https://github.com/Silver978/egUnityRepo.git
</code></pre>
<p>Is it possible to recover the lost changes anyway?</p>

**Comment by steveward** [comment, marked as answer]
<p>What you can do is create a temporary recovery branch off of the commit you made prior to reverting your changes. You can then merge that recovery branch into your <code>master</code> branch once you have verified that everything is correct. Here’s how to do that using the command line:</p>
<p><code>git checkout -b recovery d6c3e83</code></p>
<p>This creates a <code>recovery</code> branch off of the SHA of your previous commit. You should see this branch in GitHub Desktop after running that command.</p>
<p>Let me know if you run into any issues with this.</p>

**Comment by silver978** [comment]
<p>You are my savior, really, thank you so much!</p>
<p>The new branch “recovery” has the correct project version I had before the revert: problem solved!</p>
<p>So, the files of the old commits are stored in some local folders (some sort of “backup”, “recovery” folders)? Because I didn’t push that commit.</p>
<p>I ask this only to understand what I did and how git works: if this is something already explained completely in the docs I’ll go straight to study it.</p>
<p>Thank you again and have a nice day! :smiley:</p>

**Comment by steveward** [comment]
<p>I’m glad to hear you were able to recover your files!</p>
<p>&gt;So, the files of the old commits are stored in some local folders (some sort of “backup”, “recovery” folders)? Because I didn’t push that commit.</p>
<p>Think of a commit as a snapshot of the current state of your files that is stored in a database. Even if you make changes or delete those files from your local filesystem, that snapshot is still stored in the database, which means you can recover those changes at any time. </p>
<p>If you are interested in learning more I’d highly recommend this <a href="https://git-scm.com/book/en/v2/Getting-Started-What-is-Git%3F" rel="nofollow noopener">section of the book Pro Git.</a></p>

**Comment by script526** [comment]
<p>Thanks a lot. This solution has just recovered all my stuff that I was working on almost 2 years. But I’m still shaking.</p>

**Comment by Samsidirfan** [comment]
Help me to recovery lost files

**Comment by jfullstackdev** [comment]
and you don't actually pile up commits, make them small even in our company here, it's a practice, make that as your practice too

the tendency when there is a very big commits are losing the track, merge conflicts which you don't want to happen really as it can lose changes too, it's a nightmare, and unintended bugs since you are changing a lot and not noticing it's affecting other parts of the codebase too, just one mistake and you just ruined the entire codebase
